cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -503,27 +503,103 @@ impl MachInstEmit for Inst {
                     sink,
                 );
             }
+            &Inst::DivArith32 { rd, rs1, rs2 } => {
+                let rs1 = allocs.next(rs1);
+                let rs2 = allocs.next(rs2);
+                debug_assert_eq!(rs1, e0());
+                debug_assert_eq!(rs2, b0());
+                let rd = allocs.next_writable(rd);
+                // E / B => A
+                put_string("0 => D\n", sink);
+                put_string("$${var _divArith = E / B}\n", sink);
+                put_string("$${var _remArith = E % B}\n", sink);
+                put_string("${_divArith} => A\n", sink);
+                put_string("${_remArith} => C\n", sink);
+                put_string("E:ARITH\n", sink);
+
+                // A *= 2 ** 32
+                put_string("4294967296n => B\n", sink);
+                put_string("0 => C\n", sink);
+                put_string("$${var _mulArith = A * B}\n", sink);
+                // Intentionnaly ignore 64-bit overflow here 
+                // (don't write _mulArith / 2 ** 64 => D)
+                // because if it is non-zero something is not OK
+                put_string("${_mulArith} => A :ARITH\n", sink);
+            }
             &Inst::DivArith { rd, rs1, rs2 } => {
                 let rs1 = allocs.next(rs1);
                 let rs2 = allocs.next(rs2);
                 debug_assert_eq!(rs1, e0());
                 debug_assert_eq!(rs2, b0());
                 let rd = allocs.next_writable(rd);
-                // Same as in MulArith about D register
                 put_string("0 => D\n", sink);
                 put_string("$${var _divArith = E / B}\n", sink);
                 put_string("$${var _remArith = E % B}\n", sink);
                 put_string("${_divArith} => A\n", sink);
                 put_string("${_remArith} => C\n", sink);
                 put_string("E:ARITH\n", sink);
             }
+            // Rem32 is quite difficult opcode. Here we need calculate
+            // ((op1 / 2**32) % (op2 / 2**32)) * 2**32
```

Comment:
Also, there are many different ways to implement this opcode, using different register as inputs and outputs for different operations. And I believe this one is not the best. But I think anyway this operation is quite difficult and optimization will be relatively small, measuring in lines of zkasm code. So I suggest keep this version for now as working version, and in future think about more optimal implementation, if it will be necessary